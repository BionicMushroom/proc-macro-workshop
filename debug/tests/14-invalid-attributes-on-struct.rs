use derive_debug::CustomDebug;

#[derive(CustomDebug)]
#[debug]
struct Test0;

#[derive(CustomDebug)]
#[debug = "0b{:08b}"]
struct Test1;

#[derive(CustomDebug)]
#[debug(bnd = "T: ::core::fmt::Debug")]
struct Test2<T> {
    field0: T,
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
#[debug(bound = "T: ::core::fmt::Debug")]
struct Test3<T> {
    field0: T,
}

#[derive(CustomDebug)]
struct Test4 {
    #[debug = "0b{:08b}"]
    #[debug = "0b{:08b}"]
    field0: u8,
}

#[derive(CustomDebug)]
struct Test5<T> {
    #[debug(bound = "T: ::core::fmt::Debug")]
    #[debug(bound = "T: ::core::fmt::Debug")]
    field0: T,
}

#[derive(CustomDebug)]
struct Test6<T> {
    #[debug(bound = "T: ::core::fmt::Debug")]
    #[debug = "0b{:08b}"]
    #[debug(bound = "T: ::core::fmt::Debug")]
    field0: T,
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
struct Test7<T> {
    #[debug(bound = "T: ::core::fmt::Debug")]
    field0: T,
}

#[derive(CustomDebug)]
struct Test8<T> {
    #[debug(bound = "T: ::core::fmt::Debug")]
    field0: T,

    #[debug(bound = "T: ::core::fmt::Debug")]
    field1: T,
}

#[derive(CustomDebug)]
struct Test9<T> {
    #[debug]
    field0: T,
}

#[derive(CustomDebug)]
#[debug(bound = "T: ::core::fmt::Debug")]
struct Test10<T> {
    #[debug]
    field0: T,
}

fn main() {}
