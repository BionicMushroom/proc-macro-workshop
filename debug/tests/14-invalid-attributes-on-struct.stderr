error: expected `debug(bound = "...")` bound attribute
 --> tests/14-invalid-attributes-on-struct.rs:4:3
  |
4 | #[debug]
  |   ^^^^^

error: `debug = "..."` format attribute is allowed only on struct and enum fields
 --> tests/14-invalid-attributes-on-struct.rs:8:1
  |
8 | #[debug = "0b{:08b}"]
  | ^^^^^^^^^^^^^^^^^^^^^

error: expected `debug(bound = "...")` bound attribute
  --> tests/14-invalid-attributes-on-struct.rs:12:3
   |
12 | #[debug(bnd = "T: ::core::fmt::Debug")]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same struct
  --> tests/14-invalid-attributes-on-struct.rs:19:1
   |
19 | #[debug(bound = "T: ::core::fmt::Debug")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug = "..."` format attributes on the same struct field
  --> tests/14-invalid-attributes-on-struct.rs:27:5
   |
27 |     #[debug = "0b{:08b}"]
   |     ^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same struct field
  --> tests/14-invalid-attributes-on-struct.rs:34:5
   |
34 |     #[debug(bound = "T: ::core::fmt::Debug")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: multiple `debug(bound = "...")` bound attributes on the same struct field
  --> tests/14-invalid-attributes-on-struct.rs:42:5
   |
42 |     #[debug(bound = "T: ::core::fmt::Debug")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug(bound = "...")` bound attribute is not allowed on struct fields if already specified on the struct itself
  --> tests/14-invalid-attributes-on-struct.rs:49:5
   |
49 |     #[debug(bound = "T: ::core::fmt::Debug")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `debug(bound = "...")` bound attribute was already specified on this struct field type
  --> tests/14-invalid-attributes-on-struct.rs:58:5
   |
58 |     #[debug(bound = "T: ::core::fmt::Debug")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expected either `debug = "..."` format attribute or `debug(bound = "...")` bound attribute
  --> tests/14-invalid-attributes-on-struct.rs:64:7
   |
64 |     #[debug]
   |       ^^^^^

error: expected `debug = "..."` format attribute
  --> tests/14-invalid-attributes-on-struct.rs:71:7
   |
71 |     #[debug]
   |       ^^^^^
